### PR TITLE
Provide high_precision argument for predict_proba method

### LINKED ISSUE #1640: NaN probability output from LabelModel.predict_proba()
## Issue description

I encounter NaN label probabilities in the output of the LabelModel during binary labeling task when I use many labeling functions with many matches per instance. I was able to reproduce the bug with this (not entirely realistic, but valid) example.

## Code example/repro steps

```
import numpy as np
import pandas as pd

from snorkel.labeling.model import LabelModel

n_lfs = 500

# define lf-to-class mapping, i.e. with what class does an LF label instances
# shape (1 x number_of_LFs), values from [0,1] for binary task
np.random.seed(1234)
lf_assignments = np.random.binomial(n=1, p=1, size=(1, n_lfs))

# create 100 dense instances where all LFs match for every instance
dense = np.tile(A=lf_assignments, reps=(100, 1))

# create sparse instances with a single LF match for every LF
sparse = np.ones((n_lfs, n_lfs)) * -1
np.fill_diagonal(sparse, lf_assignments)

# create LabelModel input from two blocks 
L = np.concatenate([
    dense, 
    sparse
])

model= LabelModel()
model.fit(L, n_epochs=10, lr=0.0001, log_freq=5, seed=1234)

# predict probabilities; a warning is raised here (see output below)
label_probs = model.predict_proba(L)
print(np.isnan(label_probs).any())  # returns True
```
This part is executed without errors, but the `label_probs` contains NaNs. This is the output of the code snippet, including the warning:

```
2021-03-31 13:04:15,298 root         INFO     Computing O...
2021-03-31 13:04:15,743 root         INFO     Estimating \mu...
2021-03-31 13:04:15,790 root         INFO     [0 epochs]: TRAIN:[loss=6856.269]
2021-03-31 13:04:16,003 root         INFO     [5 epochs]: TRAIN:[loss=6812.121]
2021-03-31 13:04:16,249 root         INFO     Finished Training
/mounts/Users/student/speranskaya/.local/lib/python3.8/site-packages/snorkel/labeling/model/label_model.py:424: RuntimeWarning: invalid value encountered in true_divide
  return X / Z
True
```

Additional investigation of the `label_probs` shows, that the NaNs are assigned to the dense block of the input, i.e. only for the instances with many LF matches: 

```
df = pd.DataFrame({
    # count how many LFs matched for an instance
    "match_count": (L != -1).sum(1).flatten().tolist(), 
    
    # boolean indicator whether instance got a NaN label probability
    "has_nan": np.isnan(label_probs).any(1).flatten().tolist()})
    
# whole dense part recieves NaN labels (100 instances)
print(f"There are {df.has_nan.value_counts()[True]} NaN probability outputs.")  # there are 100 NaN outputs
print("Mean amount of LF matches for instances with NaN output label probability: "
      f"{df.loc[df.has_nan, 'match_count'].mean()}")  # all of them have 500 LF matches => dense block
```

Output:
```
There are 100 NaN probability outputs.
Mean amount of LF matches for instances with NaN output label probability: 500.0
```

## Expected behavior
An error or a warning if the input and/or the parameters of LabelModel do not meet any mandatory restrictions, or valid float output probabilities otherwise.

## System info

* Snorkel installed via pip
* OS: Linux
* Python version: 3.8.5
* Snorkel version: 0.9.7 
* Versions of any other relevant libraries: numpy 1.17.4 

## Additional context
1. I additionally investigated the problem by adding the following lines to [label_model.py#L417](https://github.com/snorkel-team/snorkel/blob/b3b0669f716a7b3ed6cd573b57f3f8e12bcd495a/snorkel/labeling/model/label_model.py#L417)

```
# Note: We omit abstains, effectively assuming uniform distribution here
X = np.exp(L_aug @ np.diag(jtm) @ np.log(mu) + np.log(self.p))
print(f"X has NaN values: {np.isnan(X).any()}")
Z = np.tile(X.sum(axis=1).reshape(-1, 1), self.cardinality)
print(f"X.sum() has NaN values: {np.isnan(X.sum(axis=1)).any()}")
print(f"Z has NaN values: {np.isnan(Z).any()}")
print(f"Z has zero values: {(Z == 0).any()}")
```

This outputs the following, indicating that the NaNs are probably the result of a zero division:

```
X has NaN values: False
X.sum() has NaN values: False
Z has NaN values: False
Z has zero values: True
```

2. Initially, this problem arised with even more LFs (~6500), but much lower "match density". Instances with ~300 LF matches already received a NaN probability. The loss in this setting is also much lower during fit, since the LFs are not random.

### PR DESCRIPTION
**Pull request by simonzli**
## Description of proposed changes
This PR adds a new `high_precision` argument to the `predict_proba` method in `LabelModel`.

## Related issue(s)
Fixes #1640

## Test plan
Repeat the steps as described in #1640. When setting `high_precision=False`, the output is `True` (there are NaN values). When setting `high_precision` to `True`, the output is `False`.

## Checklist

Need help on these? Just ask!

* [x] I have read the **CONTRIBUTING** document.
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [ ] I have run `tox -e complex` and/or `tox -e spark` if appropriate.
* [x] All new and existing tests passed.

### CODE REVIEW

**Comment by rsmith49** [comment]
Hi @simonzli ! Sorry for the late reply. We recently went through some CI upgrades for the checks to become more streamlined. However, because of this, existing PRs such as this one have the old CI checks still attached.

Would you mind rebasing to the new master branch and pushing this PR from there? You can accomplish that by

    git pull
    git rebase master
    git push --force

Thanks!

**Comment by rsmith49** [comment]
Hey @simonzli, checking in again. If you could run
```
git pull
git rebase master
git push --force
```

to get the CI working on this branch, I can go ahead and approve and merge it once it passes tests!
